### PR TITLE
Type: use flat_set instead of vector in ptref

### PR DESCRIPTION
**Pull request by antoine-de**
since the vector<idx> were implicitly sorted, we use a better container
for this

We need to check all `TODO: use bulk insert` to be sure that were doing the right stuff

[Review on Reviewable](https://reviewable.io/reviews/canaltp/navitia/1448)

### CODE REVIEW

File: source/ptreferential/ptreferential.cpp [hunk 1]
```diff
@@ -176,11 +175,11 @@ struct GetterType<ObjFactory<T>> {
 };
 
 template<typename T, typename C>
-std::vector<idx_t> filtered_indexes(const T& data, const C& clause) {
-    std::vector<idx_t> result;
+boost::container::flat_set<idx_t> filtered_indexes(const T& data, const C& clause) {
```
Comment (pbougue) [inline source/ptreferential/ptreferential.cpp]: You can use `Indexes` here

Comment (xlqian) [inline source/ptreferential/ptreferential.cpp]: +1

File: source/ptreferential/ptreferential.cpp [hunk 2]
```diff
@@ -201,22 +200,22 @@ inline constexpr std::false_type has_string_lookup(const Container&, ...) {
...
 
 template<typename T, typename Container>
-std::vector<idx_t> filter_by_uri(const Container& c, const std::string& uri, std::true_type) {
+Indexes filter_by_uri(const Container& c, const std::string& uri, std::true_type) {
     // for associative containers we can do a better search
     auto elt = find_or_default(uri, c);
-    if (elt) { return {elt->idx}; }
-    return {};
+    if (elt) { return Indexes({elt->idx}); }
+    return Indexes{};
```
Comment (xlqian) [inline source/ptreferential/ptreferential.cpp:213]: minor: old code should work too.. no?

File: source/ptreferential/ptreferential.cpp [hunk 3]
```diff
@@ -350,59 +349,43 @@ std::vector<Filter> parse(std::string request){
...
+Indexes get_intersection(Indexes& idxs1, Indexes& idxs2) {
+   Indexes tmp_indexes;
+   std::insert_iterator<Indexes> it(tmp_indexes, std::begin(tmp_indexes));
+   std::set_intersection(std::begin(idxs1), std::end(idxs1), std::begin(idxs2), std::end(idxs2), it);
    return tmp_indexes;
 }
 
-std::vector<idx_t> manage_odt_level(const std::vector<type::idx_t>& final_indexes,
+Indexes manage_odt_level(const Indexes& final_indexes,
                                           const navitia::type::Type_e requested_type,
```
Comment (pbougue) [inline source/ptreferential/ptreferential.cpp]: very minor: indent

File: source/ptreferential/ptreferential.cpp [hunk 4]
```diff
@@ -271,9 +270,9 @@ std::vector<idx_t> get_indexes(Filter filter,  Type_e requested_type, const Data
             case Type_e::POI: tmp = d.geo_ref->poi_proximity_list.find_within(coord, distance);break;
             default: throw ptref_error("The requested object can not be used a DWITHIN clause");
             }
-            for(auto pair : tmp) {
-                indexes.push_back(pair.first);
-            }
+            std::vector<idx_t> tmp_idx;
+            for (const auto& p : tmp) { tmp_idx.push_back(p.first); }
+            indexes.insert(tmp_idx.begin(), tmp_idx.end());
```
Comment (xlqian) [inline source/ptreferential/ptreferential.cpp:275]: why don't you push directly into `indexes`?

File: source/ptreferential/ptreferential.cpp [hunk 5]
```diff
@@ -350,59 +349,43 @@ std::vector<Filter> parse(std::string request){
...
-                                 std::vector<type::idx_t>& list_idx2){
-   std::vector<type::idx_t>::iterator new_end_1 = sort_and_ge_new_end(list_idx1);
-   std::vector<type::idx_t>::iterator new_end_2 = sort_and_ge_new_end(list_idx2);
-
-   std::vector<idx_t> tmp_indexes;
-   std::back_insert_iterator< std::vector<idx_t> > it(tmp_indexes);
-   std::set_difference(list_idx1.begin(), new_end_1,
-           list_idx2.begin(), new_end_2, it);
-   return tmp_indexes;
+Indexes get_difference(Indexes& idxs1, Indexes& idxs2) {
```
Comment (xlqian) [inline source/ptreferential/ptreferential.cpp]: why it is not `const`?

Same for L359

File: source/ptreferential/ptreferential.cpp (hunk 2 above)
Comment (antoine-de) [inline source/ptreferential/ptreferential.cpp:213]: The flat_set constructor needs to be explicit

File: source/ptreferential/ptreferential.cpp (hunk 4 above)
Comment (antoine-de) [inline source/ptreferential/ptreferential.cpp:275]: It's to to a bulk insert

**Comment by xlqian** [comment]
ok for me

File: source/ptreferential/ptreferential.cpp (hunk 4 above)
Comment (xlqian) [inline source/ptreferential/ptreferential.cpp:275]: ok

**Comment by TeXitoi** [comment]
OK for me.

**Pull request merged**